File: Medium/2026-05-07-3660-Jump-Game-IX/solution.java
import java.util.Arrays;
import java.util.Stack;

public class Solution {
    public int[] maxJump(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        Arrays.fill(ans, Integer.MIN_VALUE);
        
        // Forward pass
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i]) {
                ans[stack.pop()] = Math.max(ans[stack.peek()], nums[i]);
            }
            stack.push(i);
        }
        
        // Clear stack for backward pass
        stack.clear();
        
        // Backward pass
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && nums[stack.peek()] > nums[i]) {
                ans[stack.pop()] = Math.max(ans[stack.peek()], nums[i]);
            }
            stack.push(i);
        }
        
        // Fill in the maximum values for each index
        for (int i = 0; i < n; i++) {
            ans[i] = Math.max(ans[i], nums[i]);
        }
        
        return ans;
    }
}

File: Medium/2026-05-07-3660-Jump-Game-IX/solution.js
var maxJump = function(nums) {
    const n = nums.length;
    const ans = new Array(n).fill(0);
    const stack = [];

    for (let i = 0; i < n; i++) {
        while (stack.length && nums[stack[stack.length - 1]] < nums[i]) {
            ans[stack.pop()] = Math.max(ans[stack[stack.length - 1]] || 0, nums[i]);
        }
        stack.push(i);
    }

    stack.length = 0; // Clear the stack for the next pass

    for (let i = n - 1; i >= 0; i--) {
        while (stack.length && nums[stack[stack.length - 1]] > nums[i]) {
            ans[stack.pop()] = Math.max(ans[stack[stack.length - 1]] || 0, nums[i]);
        }
        stack.push(i);
    }

    for (let i = 0; i < n; i++) {
        ans[i] = Math.max(ans[i], nums[i]);
    }

    return ans;
};